Nvenc encoder #259 Support (rebased)

sgothel opened the pull request:
Rebase and fix of pull request #983 of issue #259 
Still need to apply suggestions of @sr55 
(github doesn't allow to re-open a rebased pull request)

sgothel committed:
NVENC: Adding ffmpeg nvenc.c patch to add verbosity on parameters

sgothel committed:
FFMPEG #974: Use latest FFMPEG 3.4.1 (1/2)

Patch 1/2 for https://github.com/HandBrake/HandBrake/issues/974
moving to FFMPEG 3.4.1 from LIBAV 12.2.

The re-ordering of to be linked modules was required to solve
statically linked dependencies. See libhb/module.defs etc.

sgothel committed:
FFMPEG #974: Use latest FFMPEG 3.4.1 (2/2)

Patch 2/2 for https://github.com/HandBrake/HandBrake/issues/974
moving to FFMPEG 3.4.1 from LIBAV 12.2.

All patches have been moved to subfolder 'old' since they do not apply cleanly anymore.
Work has to be performed to validate whether patches are still required.

MatthewLazarow committed:
Initial nvenc h264/hevc encoder patch

sgothel committed:
NVENC: Remaining fixes for 1st GNU/Linux GTK working draft

- Fix HB_VCODEC_FFMPEG_MASK covered values to lie within mask
- Fix HB_VCODEC_FFMPEG_H264 usage (gtk version missed a rename)
- Fix FFMPEG hwaccel build is required for Nvenc

Basically working:
  - h264_nvenc
  - hevc_nvenc
    Possible optimization:
  - Even in constant quality mode, a user ceiling bit_rate might be desirable,
    currently only the maximum is used like in VP8/9: w x h x fps

  Remaining issues:
    1) Muxer issue with h264_nvenc + MKV, ffprobe shows:

    [h264 @ 0x55ae4dedacc0] sps_id 32 out of range
    [h264 @ 0x55ae4dedacc0] Invalid nal size 1677734572
    [h264 @ 0x55ae4dedacc0] missing picture in access unit with size 40644

     Notable: This does not happen with hevc_nvenc

  Tested:
    - GNU/Linux Debian 9
    - GeForce GTX 950
    - BD 1080p source, cropped: Around 70fps h264 + hevc

sgothel committed:
NVENC: Fix VBR CQ Settings to be more adaptive

Adding 'rc-lookahead 32, init_qp 1, spatial_aq 1, aq-strength 8' for h264 and hevc
in case of VBR CQ mode, see
  <https://devtalk.nvidia.com/default/topic/1013152/video-technologies/crf-for-nvenc/post/5179877/#5179877>
and NVENC_VideoEncoder_API_ProgGuide.pdf v08 Chapter 8 p27, as well as Table 1 p33.

sgothel committed:
NVENC: Refining ffmpeg nvenc.c patch to refine verbosity on parameters

1st patch was ab0cca4b53fcabb1ab7fe1cf01306f052cb4b89d
NVENC: Adding ffmpeg nvenc.c patch to add verbosity on parameters

sgothel committed:
muxavformat: more verbose error msg

sgothel committed:
NVENC: Fix VBR CQ Settings to be more adaptive (2)

Initial VBR CQ patch 037de2cf670092ab0881e06a43df383be1628de0
'NVENC: Fix VBR CQ Settings to be more adaptive'
caused encoding errors w/ max lookahead 32.
Now using lookahead 16 only plus the driver settings default of aq-strength.

This fixes encoding/muxing using nvenc and also removes related av-sync issues
on certain input streams.

sgothel committed:
Fix hb_job_t path in comment

@sr55 'Something odd happened with this comment. It was incorrect. HandBrake.Interop is gone. Correct path is: HandBrake.ApplicationServices\Interop\HbLib'
https://github.com/HandBrake/HandBrake/pull/983#discussion_r149801776

sgothel commented:
@sr55  on Nov 8 Owner https://github.com/HandBrake/HandBrake/pull/983#discussion_r149801432

> We should be able to bring these options though the extra arguments string, 
> split that and feed those to the encoder. 
> Having full control over the encoder is defiantly preferable.

Agreed. We shall have the currently used params as a default, allowing them to be overwritten.
If you can show me how to access custom parameters I may can serve this handling.

Currently nvenc uses 2 sets of default parameters: 'VBR CQ' and 'fixed bitrate'

bradleysepos commented:
You can push a rebased branch using `git push --force ...`

jstebbins commented:
> You can push a rebased branch using git push --force ...

Just be very careful that you are pushing to the correct branch :trollface: 
That said, I use --force *all the time* when doing PRs.  It's the easiest way to stay current as master changes under you.

sgothel committed:
NVENC: Fix HB_VCODEC_FFMPEG_H264 -> HB_VCODEC_FFMPEG_H264_{NVENC,VAAPI}

Adding the specific FFMPEG encoder suffix allows support for more encoders like VAAPI.

Cleaned up overall HB_VCODED_* to 32bit signed integer space within 0x7FFFFFFF.

sgothel committed:
Using AV_CODEC_ID_SUBRIP instead of AV_CODEC_ID_SRT (libav->ffmpeg)

This fixes issue https://trac.ffmpeg.org/ticket/6304
See https://github.com/HandBrake/HandBrake/pull/981#issuecomment-347364763
See https://github.com/HandBrake/HandBrake/issues/974#issuecomment-353986772

sgothel committed:
Merge branch 'ffmpeg' into nvenc-encoder

sgothel committed:
Merge branch 'master' into ffmpeg

sgothel committed:
Merge branch 'ffmpeg' into nvenc-encoder

sgothel committed:
FFMPEG: Enable A01-matroskaenc-pgs-duration patch

See http://ffmpeg.org/pipermail/ffmpeg-devel/2017-November/219642.html

sgothel committed:
Merge branch 'ffmpeg' into nvenc-encoder

sgothel committed:
NVENC: Utilize HB_VCODEC_FFMPEG_*_NVENC where applicable

sgothel committed:
NVENC: Properly test whether encoder is available on platform

Proper availability tests requires to open the codec's context,
only if successful we can deem it available on the runtime platform.

sgothel committed:
NVENC: Remove early added unused WORK ids

sgothel committed:
FFMPEG: Use avcodec_free_context(..) instead of deprecated leaking avcodec_close(..)

Hence rename hb_avcodec_close -> hb_avcodec_free_context and pass the required ptr-ptr.
avcodec_free_context(..) ensures releasing of all resources attached to the context.

sgothel committed:
NVENC: HB_VCODEC_{H264,H265}_MASK bits not unique, use matching macro def instead

Also add HB_VCODEC_FFMPEG_{VP8,VP9}_VAAPI for future use.

sgothel committed:
NVENC: Remove redundant ffmpeg config args

sgothel committed:
FFMPEG decavcodec: Consider returned error from av_buffersrc_add_frame(..)

sgothel committed:
Merge branch 'ffmpeg' into nvenc-encoder

sgothel committed:
Merge branch 'master' into ffmpeg

sgothel committed:
Merge branch 'ffmpeg' into nvenc-encoder

sgothel commented:
Added commits in nvenc-encoder branch, not showing here (yet). Hmm.

Dracozny commented:
Re-encoding videos with subtitles results in failure: Feature not enabled.

sr55 commented:
@sgothel  Now that we have ffmpeg 4.0 and some changes for AMD VCE in, do you plan to re-base this?

If not, let me know and I might take it over to get it over the line.

sr55 commented:
Closing.  I'll find some time to take this over unless anyone pings me.
